Fix #33448: Format WP column "BCF snapshot" as string

wielinde opened the pull request:
Render an "X" if there is at least one viewpoint for the associated BCF issue.

https://community.openproject.org/wp/33448

wielinde committed:
Fix #33448: Format WP column "BCF snapshot" as string

Render an "X" if there is at least one viewpoint for the
associated BCF issue.

https://community.openproject.org/wp/33448

oliverguenther commented on app/models/work_package/exporter/formatters.rb:
```diff
@@ -1,19 +1,19 @@
 module WorkPackage::Exporter
   module Formatters
+    @@formatters = %i[default costs estimated_hours].map do |key|
```
I prefer the memoized style with an instance-level variable. Class variables are generally frowned upon due to the inheritance problems. Even though this module is likely not being subclassed, why introduce a class variable?

wielinde replied:
Yes, I already asked Jens how to improve that because I knew that it "smelled". I'll improve that. Thanks.

wielinde replied:
changed it.

oliverguenther commented on app/models/work_package/pdf_export/work_package_list_to_pdf.rb:
```diff
@@ -317,4 +317,10 @@ def batch_supported?
     end
   end
+
+  def make_bcf_thumbnail_field_value(work_package)
```
This method is added but never used?

wielinde replied:
removed it.

oliverguenther commented on app/models/work_package/exporter/formatters.rb:
```diff
@@ -1,19 +1,19 @@
-          Kernel.const_get("WorkPackage::Exporter::Formatters::#{key.to_s.camelize}")
-        end
-      end
+      @@formatters
```
As stated above, I'd prefer to revert these changes into the instance variable pattern

oliverguenther reviewed (changes requested):
Some minor code remarks

wielinde committed:
Post review modifications

- Don't use class variables.
- Use memoized class instance variables.
- Fix class reloading in development environment.

wielinde commented:
**Pre review modification goals:**

- Make registering of formatters work. That is why I introduced the class variables.... bad idea.

**Post review modifications**
    
- Don't use class variables.
- Use memoized class instance variables.
- Fix class reloading in development environment.

wielinde committed:
Satisfy rubocop

wielinde committed:
fixing logic

wielinde commented:
I don't see the point of the last remaining rubocop errors. `apply?` is a method of the code that gets tested. It is not code testing it.
